### PR TITLE
Add headers to BaseClient

### LINKED ISSUE #149: Set Headers like Cookies in Client session for all requests
Assuming one can use `Client`/`AsyncClient` as a drop-in replacement for `requests.Session`, I would like to be able to set headers (and cookies) to be sent for each request, for example for token-based authentication.

With `requests`, I can do:
```python
header_dict = {"X-Auth-Token": "FooBarBazToken"}
session = requests.Session()
session.headers.update(header_dict)
```

`cookies` can already be set on the `httpx.Client`. I would like to be able to also do that with `headers`:
```python
>>> headers = Headers({"X-Auth-Token": "FooBarBazToken", "User-Agent": "python-myclient/0.2.1"})
>>> session = Client(base_url="https://httpbin.org", headers=headers)

>>> get_headers = session.get("/headers")

>>> get_headers.json()
{'headers': {'Accept': '*/*',
  'Accept-Encoding': 'gzip, deflate',
  'Cookie': 'MyCookie=MyValue',
  'Host': 'httpbin.org',
  'User-Agent': 'python-myclient/0.2.1',
  'X-Auth-Token': 'FooBarBazToken'}}
```

### PR DESCRIPTION
**Pull request by StephenBrown2**
This PR resolves #149, by adding `headers` as an arg to the `BaseClient` class, as with `cookies`, and also adds a `merge_headers` function to... merge the headers. I started by copying `merge_cookies`, but got a `NoneType is not iterable` error when trying to use it, so I changed the conditional to ensure that both the passed `headers` and `self.headers` were set, and returning one or the other in the case that they are not. This will still return `None` if neither are set.

I haven't added test cases, yet, but I did test against httpbin.org. I could check out adding something based on the existing [cookie tests](https://github.com/encode/httpx/blob/master/tests/client/test_cookies.py), but wanted to get some feedback first.

```
Python 3.7.3 (default, Jun 24 2019, 04:54:02) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.6.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from httpx import Headers, Client, AsyncClient                                                                                                                                                               

In [2]: headers = Headers({"X-Auth-Token": "FooBarBazToken", "Fake-Agent": "python-myclient/0.2.1"}) 
   ...: session = Client(base_url="https://httpbin.org", headers=headers)                                                                                                                                        

In [3]: async_session = AsyncClient(base_url="https://httpbin.org", headers=headers)                                                                                                                                 

In [4]: sync_headers = session.get("/headers")                                                                                                                                                                       

In [5]: async_headers = await async_session.get("/headers")                                                                                                                                                          

In [6]: sync_headers.json()                                                                                                                                                                                          
Out[6]: 
{'headers': {'Accept': '*/*',
  'Accept-Encoding': 'gzip, deflate',
  'Fake-Agent': 'python-myclient/0.2.1',
  'Host': 'httpbin.org',
  'User-Agent': 'python-httpx/0.6.8',
  'X-Auth-Token': 'FooBarBazToken'}}

In [7]: async_headers.json()                                                                                                                                                                                         
Out[7]: 
{'headers': {'Accept': '*/*',
  'Accept-Encoding': 'gzip, deflate',
  'Fake-Agent': 'python-myclient/0.2.1',
  'Host': 'httpbin.org',
  'User-Agent': 'python-httpx/0.6.8',
  'X-Auth-Token': 'FooBarBazToken'}}

In [8]: headers = Headers({"X-Auth-Token": "FooBarBazToken", "User-Agent": "python-myclient/0.2.1"}) 
   ...: session_two = Client(base_url="https://httpbin.org", headers=headers)                                                                                                                                    

In [9]: async_session_two = AsyncClient(base_url="https://httpbin.org", headers=headers)                                                                                                                             

In [10]: sync_headers_two = session.get("/headers")                                                                                                                                                                  

In [11]: async_headers_two = await async_session.get("/headers")                                                                                                                                                     

In [12]: sync_headers_two.json()                                                                                                                                                                                     
Out[12]: 
{'headers': {'Accept': '*/*',
  'Accept-Encoding': 'gzip, deflate',
  'Fake-Agent': 'python-myclient/0.2.1',
  'Host': 'httpbin.org',
  'User-Agent': 'python-httpx/0.6.8',
  'X-Auth-Token': 'FooBarBazToken'}}

In [13]: async_headers_two.json()                                                                                                                                                                                    
Out[13]: 
{'headers': {'Accept': '*/*',
  'Accept-Encoding': 'gzip, deflate',
  'Fake-Agent': 'python-myclient/0.2.1',
  'Host': 'httpbin.org',
  'User-Agent': 'python-httpx/0.6.8',
  'X-Auth-Token': 'FooBarBazToken'}}

In [14]: async_session.headers.update({"Another-Header": "IsAThing"})                                                                                                                                                

In [14]: async_headers_three = await async_session.get("/headers")                                                                                                                                                   

In [15]: async_headers_three.json()                                                                                                                                                                                  
Out[15]: 
{'headers': {'Accept': '*/*',
  'Accept-Encoding': 'gzip, deflate',
  'Another-Header': 'IsAThing',
  'Fake-Agent': 'python-myclient/0.2.1',
  'Host': 'httpbin.org',
  'User-Agent': 'python-httpx/0.6.8',
  'X-Auth-Token': 'FooBarBazToken'}}

```

### CODE REVIEW

File: httpx/client.py [hunk 1]
```diff
@@ -109,6 +111,15 @@ def merge_cookies(
...
         return cookies
 
+    def merge_headers(
+        self, headers: HeaderTypes = None
+    ) -> typing.Optional[HeaderTypes]:
+        if headers and self.headers:
+            merged_headers = Headers(self.headers)
+            merged_headers.update(headers)
+            return merged_headers
+        return headers or self.headers
```
Comment (yeraydiazdiaz) [inline httpx/client.py]: The reason you're getting the `NoneType is not iterable error` is because `Headers` does not implement the `update` method, [unlike `Cookies`](https://github.com/encode/httpx/blob/master/httpx/models.py#L1110-L1113), and falls back to the `MutableDict` implementation which does not handle the `None` case. Maybe it's worth implement it if only for consistency?

File: httpx/models.py [hunk 2]
```diff
@@ -309,7 +309,7 @@ def __repr__(self) -> str:
         return f"{class_name}({query_string!r})"
 
 
-class Headers(typing.MutableMapping[str, str]):
+class Headers(typing.MutableMapping[str, str], MutableMapping):
```
Comment (yeraydiazdiaz) [inline httpx/models.py]: Why is this needed?

File: httpx/client.py (hunk 1 above)
Comment (StephenBrown2) [reply httpx/client.py]: Sounds good, I've implemented `update` to mimic Cookies now, and it appears to work just the same. :+1:

File: httpx/models.py (hunk 2 above)
Comment (StephenBrown2) [reply httpx/models.py]: I had added it thinking it would inherit the update function, but it turns out that's not how it worked. Removed it now.

**Comment by StephenBrown2** [comment]
Definitely, I've added the param to the api spec, but the docs could use a larger fleshing out/update most likely.

**Comment by sethmlarson** [comment]
Yeah that'll be a pre-1.0 release probably.

**Comment by sethmlarson** [comment]
Looks like CI is failing a few tests, take a look at those.

**Comment by StephenBrown2** [comment]
Good catch, forgot what object I was in, fixed now.

**Comment by StephenBrown2** [comment]
Sure thing. Do we also want tests for cookies sent from the client since it's a similar use case?

**Comment by sethmlarson** [comment]
I added a single test-case that verifies the way that conflicting headers are resolved so we maintain that property of how they've resolved. :)

**Comment by sethmlarson** [comment]
Thank you @StephenBrown2 for this great change! :)

**Comment by tomchristie** [comment]
Great stuff! Related to this - I reckon we'll probably want a `BaseClient.build_request()` method as a public interface.
